[receiver/prometheusremotewritereceiver] accept unspecified histogram

jelly-afk opened the pull request:
<!--Ex. Fixing a bug - Describe the bug and how this fixes the issue.
Ex. Adding a feature - Explain what this achieves.-->
#### Description
prometheusremote write receiver now accepts histogram with unspecified metric type.

<!-- Issue number (e.g. #1234) or full URL to issue, if applicable. -->
#### Link to tracking issue
Fixes #41840 

<!--Describe what testing was performed and which tests were added.-->
#### Testing
Added a testcase in ` TestTranslateV2` test function to test unspecified histogram.

jelly-afk committed:
accept unspecified histogram

dashpole reviewed (commented):
This will only work for CBNH, not standard "split" histograms (e.g. with _bucket _sum and _count series).  Will this actually fix your use-case?  I'm very surprised that you can get a CBNH without a histogram type...

perebaj committed:
[prometheusremotewritereceiver] concurrency bug (#43383)

<!--Ex. Fixing a bug - Describe the bug and how this fixes the issue.
Ex. Adding a feature - Explain what this achieves.-->

Fixed a concurrency bug in the Prometheus remote write receiver where
concurrent requests with identical job/instance labels would return
empty responses after the first successful request.

<!-- Issue number (e.g. #1234) or full URL to issue, if applicable. -->
Fixes #42159

<!--Describe what testing was performed and which tests were added.-->

<!--Describe the documentation added.-->

<!--Please delete paragraphs that you did not use before submitting.-->

---------

Signed-off-by: perebaj <perebaj@gmail.com>
Co-authored-by: Vihas Makwana <121151420+VihasMakwana@users.noreply.github.com>
Co-authored-by: Arthur Silva Sens <arthursens2005@gmail.com>

ArthurSens commented:
> This will only work for CBNH, not standard "split" histograms (e.g. with _bucket _sum and _count series). Will this actually fix your use-case? I'm very surprised that you can get a CBNH without a histogram type...

I think that's the common case when Prometheus doesn't have `metadata-wal-records` enabled, isn't it? Prometheus remote-write heavily depends on WAL, so if metadata is not present in WAL, then everything is sent without type.

perebaj commented on receiver/prometheusremotewritereceiver/receiver.go:
```diff
@@ -487,6 +488,11 @@ func (prw *prometheusRemoteWriteReceiver) processHistogramTimeSeries(
+				histMetric.Metadata().PutStr(prometheus.MetricMetadataTypeKey, "histogram")
+			} else {
+				histMetric.Metadata().PutStr(prometheus.MetricMetadataTypeKey, "unknown")
+			}
```
Isn't better to use switch case for this use case?

jelly-afk commented on receiver/prometheusremotewritereceiver/receiver.go:
```diff
@@ -301,7 +301,8 @@ func (prw *prometheusRemoteWriteReceiver) translateV2(_ context.Context, req *wr
 		// Handle histograms separately due to their complex mixed-schema processing
-		if ts.Metadata.Type == writev2.Metadata_METRIC_TYPE_HISTOGRAM {
+		if ts.Metadata.Type == writev2.Metadata_METRIC_TYPE_HISTOGRAM ||
+			ts.Metadata.Type == writev2.Metadata_METRIC_TYPE_UNSPECIFIED && len(ts.Histograms) > 0 {
```
Do you mean other metric types besides histograms? :thinking:
If so, those were handled in #42295

perebaj replied:
you are right. thanks!

jelly-afk committed:
add unspecified nhcb test case
